### PR TITLE
Word Embeddings Class Added

### PR DESCRIPTION
**Pull request by sprintyaf**
**Describe your change:**
#2187 
I added Word Embeddings Class described in the above-mentioned issue.

* [x] Add an algorithm?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: natural_language_processing/word_embeddings/usage.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+word2 = 'unify'
+score = model.similarity(word1, word2)
+print('similarity between {} and {} is {}'.format(word1, word2, score))
+
+# word analogies
+x1 = 'boy'
+x2 = 'prince'
+y1 = 'girl'
+y2 = model.analogy(x1, x2, y1)
+print('{} <--> {}  |  {} <--> {}'.format(x1, x2, y1, y2))
```
Comment (cclauss) [inline natural_language_processing/word_embeddings/usage.py:27]:
```suggestion
print(f'{x1} <--> {x2}  |  {y1} <--> {y2}')
```
Please use f-strings as discussed in CONTRIBUTING.md

File: natural_language_processing/word_embeddings/usage.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+# closest words
+word = 'request'
+close_words = model.closest_words(word)
+print(close_words)
```
Comment (cclauss) [inline natural_language_processing/word_embeddings/usage.py:32]:
```suggestion
print(model.closest_words(word='request'))
```

File: natural_language_processing/word_embeddings/usage.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+print('similarity between {} and {} is {}'.format(word1, word2, score))
+
+# word analogies
+x1 = 'boy'
+x2 = 'prince'
+y1 = 'girl'
+y2 = model.analogy(x1, x2, y1)
+print('{} <--> {}  |  {} <--> {}'.format(x1, x2, y1, y2))
+
+# closest words
```
Comment (cclauss) [inline natural_language_processing/word_embeddings/usage.py:29]:
```suggestion
```
Good method names make this comment redundant.

File: natural_language_processing/word_embeddings/usage.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+word = 'request'
+close_words = model.closest_words(word)
+print(close_words)
+
+# plot closest words
+word = 'request'
+model.plot_closest_words(word, plot=False) # download the plot
```
Comment (cclauss) [inline natural_language_processing/word_embeddings/usage.py:36]:
```suggestion
model.plot_closest_words(word='request', plot=False)  # download the plot
```
Two spaces between the code and the comment

File: natural_language_processing/word_embeddings/usage.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+else:
+    model = WordVectors.train(corpus_dir)
+    model.save()
+
+
+# similarity between two words
+word1 = 'merge'
+word2 = 'unify'
+score = model.similarity(word1, word2)
+print('similarity between {} and {} is {}'.format(word1, word2, score))
```
Comment (cclauss) [inline natural_language_processing/word_embeddings/usage.py:20]: f-string

File: natural_language_processing/word_embeddings/word_embeddings.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+stop_words = set(stopwords.words('english'))
+
+"""
+This is the class of Word Embeddings.
+Model is trained with word2vec algorithm (https://youtu.be/c3yRH0XZN2g)
+"""
+
+
+class WordVectors(object):
+    def __init__(self, model):
```
Comment (cclauss) [inline natural_language_processing/word_embeddings/word_embeddings.py:19]: type hints and doctests?  Please read CONTRIBUTING.md.

File: natural_language_processing/word_embeddings/word_embeddings.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+        """
+        Init function which takes in a trained model
+        """
+        self._model = model
+        self._wv = model.wv
+        self._vocab = model.wv.vocab
+
+    
+    def analogy(self, x1, x2, y1):
```
Comment (cclauss) [inline natural_language_processing/word_embeddings/word_embeddings.py:28]: type hints and doctests?  Please read CONTRIBUTING.md.

File: natural_language_processing/word_embeddings/word_embeddings.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+        Tries to find a word which has a similar relation to y1, as x2 has to x1
+        """
+        x1, x2, y1 = x1.lower(), x2.lower(), y1.lower()
+        assert self.is_in_vocab(x1) and self.is_in_vocab(x2) and self.is_in_vocab(y1), \
+                                        'every word must be in the vocabulary'
+        result = self._wv.most_similar(positive=[y1, x2], negative=[x1])[0][0]
+        return result
+
+    
+    def n_similarity(self, list1, list2):
```
Comment (cclauss) [inline natural_language_processing/word_embeddings/word_embeddings.py:39]: type hints and doctests?  Please read CONTRIBUTING.md.

File: natural_language_processing/word_embeddings/word_embeddings.py [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+        Returns similarity score between two lists of words
+        """
+        list1 = [word.lower() for word in list1 if self.is_in_vocab(word)]
+        list2 = [word.lower() for word in list2 if self.is_in_vocab(word)]
+        assert list1 and list2, 'at least one word from each list must be in the vocabulary'
+        score = self._wv.n_similarity(list1, list2)
+        return score
+
+    
+    def similarity(self, w1, w2):
```
Comment (cclauss) [inline natural_language_processing/word_embeddings/word_embeddings.py:50]: type hints and doctests?  Please read CONTRIBUTING.md. ...

File: natural_language_processing/word_embeddings/word_embeddings.py [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+        documents = []
+        full_path = os.path.join(file_dir, '*')
+        print('parsing documents')
+        for filename in glob.glob(full_path):
+            try:
+                with open(os.path.join(os.getcwd(), filename), encoding = "ISO-8859-1") as f:
+                    text = f.read()
+                parsed = WordVectors._parse_document(text)
+                documents.append(parsed)
+            except:
```
Comment (cclauss) [inline natural_language_processing/word_embeddings/word_embeddings.py]: Bare excepts are not allowed.  See PEP8 and https://realpython.com/the-most-diabolical-python-antipattern

File: natural_language_processing/word_embeddings/word_embeddings.py [hunk 11]
```diff
@@ -24,18 +24,17 @@ def __init__(self, model):
...
-    
+
     def analogy(self, x1, x2, y1):
         """
             Tries to find a word which has a similar relation to y1, as x2 has to x1
         """
         x1, x2, y1 = x1.lower(), x2.lower(), y1.lower()
-        assert self.is_in_vocab(x1) and self.is_in_vocab(x2) and self.is_in_vocab(y1), \
-                                        'every word must be in the vocabulary'
+        assert self.is_in_vocab(x1) and self.is_in_vocab(x2) and self.is_in_vocab(y1), 'every word must be in the vocabulary'
```
Comment (cclauss) [inline natural_language_processing/word_embeddings/word_embeddings.py:33]:
```suggestion
        assert self.is_in_vocab(x1) and self.is_in_vocab(x2) and self.is_in_vocab(y1), (
            'every word must be in the vocabulary')
```

**Comment by cclauss** [comment]
Travis CI says:

$ __flake8 --ignore=E203,W503 --max-complexity=25 --max-line-length=88 --statistics --count .__
```
./natural_language_processing/word_embeddings/usage.py:36:43: E261 at least two spaces before inline comment
./natural_language_processing/word_embeddings/word_embeddings.py:28:5: E303 too many blank lines (2)
./natural_language_processing/word_embeddings/word_embeddings.py:33:89: E501 line too long (125 > 88 characters)
./natural_language_processing/word_embeddings/word_embeddings.py:38:5: E303 too many blank lines (2)
./natural_language_processing/word_embeddings/word_embeddings.py:44:89: E501 line too long (92 > 88 characters)
./natural_language_processing/word_embeddings/word_embeddings.py:49:5: E303 too many blank lines (2)
./natural_language_processing/word_embeddings/word_embeddings.py:54:89: E501 line too long (100 > 88 characters)
./natural_language_processing/word_embeddings/word_embeddings.py:59:5: E303 too many blank lines (2)
./natural_language_processing/word_embeddings/word_embeddings.py:69:5: E303 too many blank lines (2)
./natural_language_processing/word_embeddings/word_embeddings.py:91:5: E303 too many blank lines (2)
./natural_language_processing/word_embeddings/word_embeddings.py:98:5: E303 too many blank lines (2)
./natural_language_processing/word_embeddings/word_embeddings.py:106:5: E303 too many blank lines (2)
./natural_language_processing/word_embeddings/word_embeddings.py:115:89: E501 line too long (94 > 88 characters)
./natural_language_processing/word_embeddings/word_embeddings.py:119:5: E303 too many blank lines (2)
./natural_language_processing/word_embeddings/word_embeddings.py:122:68: W291 trailing whitespace
./natural_language_processing/word_embeddings/word_embeddings.py:136:72: E251 unexpected spaces around keyword / parameter equals
./natural_language_processing/word_embeddings/word_embeddings.py:136:74: E251 unexpected spaces around keyword / parameter equals
./natural_language_processing/word_embeddings/word_embeddings.py:136:89: E501 line too long (93 > 88 characters)
./natural_language_processing/word_embeddings/word_embeddings.py:144:89: E501 line too long (118 > 88 characters)
./natural_language_processing/word_embeddings/word_embeddings.py:152:5: E303 too many blank lines (2)
2     E251 unexpected spaces around keyword / parameter equals
1     E261 at least two spaces before inline comment
10    E303 too many blank lines (2)
6     E501 line too long (125 > 88 characters)
1     W291 trailing whitespace
20
```

**Comment by cclauss** [comment]
Please avoid backslash line termination as discussed in PEP8 because a whitespace character to the right of the backslash can break the script on a change that is invisibile to the user.

**Comment by sprintyaf** [comment]
Hi @cclauss,
First of all thanks for your support.
This algorithm is not deterministic and depends on the data. That's why I can't predict what answer would it return. I don't think that doctests are possible in this case.

**Comment by cclauss** [comment]
assert 70 <= my_function() <= 100, "my_function() should return a result > 70"

**Comment by sprintyaf** [comment]
> assert 70 <= my_function() <= 100, "my_function() should return a result > 70"

Agree, but my functions return words. It can return any word. You can't test them like this.

**Comment by cclauss** [comment]
`closest_words()` can return different results with the same dataset??

File: natural_language_processing/word_embeddings/word_embeddings.py [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+
+    def analogy(self, x1: str, x2: str, y1: str) -> str:
+        """
+        Tries to find a word which has a similar relation to y1, as x2 has to x1
+        """
+        x1, x2, y1 = x1.lower(), x2.lower(), y1.lower()
+        error_msg = 'every word must be in the vocabulary'
+        assert all(self.is_in_vocab(x) for x in (x1, x2, y1)), error_msg
+        result = self._wv.most_similar(positive=[y1, x2], negative=[x1])[0][0]
+        return result
```
Comment (cclauss) [inline natural_language_processing/word_embeddings/word_embeddings.py:37]:
```suggestion
        return self._wv.most_similar(positive=[y1, x2], negative=[x1])[0][0]
```

File: natural_language_processing/word_embeddings/word_embeddings.py [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+    def n_similarity(self, list1: list, list2: list) -> float:
+        """
+        Returns similarity score between two lists of words
+        """
+        list1 = [word.lower() for word in list1 if self.is_in_vocab(word)]
+        list2 = [word.lower() for word in list2 if self.is_in_vocab(word)]
+        error_msg = 'at least one word from each list must be in the vocabulary'
+        assert list1 and list2, error_msg
+        score = self._wv.n_similarity(list1, list2)
+        return score
```
Comment (cclauss) [inline natural_language_processing/word_embeddings/word_embeddings.py:48]:
```suggestion
        return self._wv.n_similarity(list1, list2)
```

File: natural_language_processing/word_embeddings/word_embeddings.py [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+
+    def closest_words(self, word: str) -> list:
+        """
+        Returns similar words for a given word
+        """
+        word = word.lower()
+        error_msg = 'the word must be in the vocabulary'
+        assert self.is_in_vocab(word), error_msg
+        close_words = self._wv.similar_by_word(word)
+        return close_words
```
Comment (cclauss) [inline natural_language_processing/word_embeddings/word_embeddings.py:68]:
```suggestion
        return self._wv.similar_by_word(word)
```
The method name already tells us that we are returning the closest words.

File: natural_language_processing/word_embeddings/word_embeddings.py [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+        """
+        Parses give text document and returns list of tokens
+        """
+        tokens = word_tokenize(text)
+        splitted_tokens = []
+        for token in tokens:
+            splitted_tokens.extend(re.split('[^0-9a-zA-Z+.#]+', token))
+        parsed = [word.lower() for word in splitted_tokens
+                  if word and word not in stop_words]
+        return parsed
```
Comment (cclauss) [inline natural_language_processing/word_embeddings/word_embeddings.py:115]:
```suggestion
        return [word.lower() for word in splitted_tokens
                if word and word not in stop_words]
```
The method name already tells us that these are parsed.

**Comment by cclauss** [comment]
Please add tests to the deterministic functions and place a comment explaining why a doctest is impossible on the non deterministic functions.

File: natural_language_processing/word_embeddings/word_embeddings.py [hunk 16]
```diff
@@ -33,8 +37,7 @@ def analogy(self, x1: str, x2: str, y1: str) -> str:
         x1, x2, y1 = x1.lower(), x2.lower(), y1.lower()
         error_msg = 'every word must be in the vocabulary'
         assert all(self.is_in_vocab(x) for x in (x1, x2, y1)), error_msg
-        result = self._wv.most_similar(positive=[y1, x2], negative=[x1])[0][0]
-        return result
+        return self._wv.most_similar(positive=[y1, x2], negative=[x1])[0][0]
 
     def n_similarity(self, list1: list, list2: list) -> float:
```
Comment (ruppysuppy) [inline natural_language_processing/word_embeddings/word_embeddings.py:42]: Add proper type hints in the functions (see [here](https://docs.python.org/3/library/typing.html))

File: natural_language_processing/word_embeddings/word_embeddings.py [hunk 17]
```diff
@@ -54,8 +56,7 @@ def similarity(self, w1: str, w2: str) -> float:
         w1, w2 = w1.lower(), w2.lower()
         error_msg = 'both words must be in the vocabulary'
         assert self.is_in_vocab(w1) and self.is_in_vocab(w2), error_msg
```
Comment (ruppysuppy) [inline natural_language_processing/word_embeddings/word_embeddings.py:58]: If the assertions are for tests, use the doctest format (see [here](https://docs.python.org/3/library/doctest.html))

Comment (cclauss) [reply natural_language_processing/word_embeddings/word_embeddings.py:58]: These assertions are for runtime checks...  If the user gives me garbage then I should complain.  These garbage values should also be tested in the doctests.

File: natural_language_processing/word_embeddings/word_embeddings.py (hunk 16 above)
Comment (cclauss) [reply natural_language_processing/word_embeddings/word_embeddings.py:42]: __Please__ Add proper type hints in the functions...

**Comment by sprintyaf** [comment]
@cclauss @ruppysuppy 
I added type hints and doctests. For doctests I needed to upload a pretrained model. But it looks like it's not visible during testing. How could I fix this?

error

File: natural_language_processing/word_embeddings/word_embeddings.py [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+        print('parsing documents')
+        for filename in glob.glob(full_path):
+            try:
+                with open(os.path.join(os.getcwd(), filename),
+                          encoding="ISO-8859-1") as f:
+                    text = f.read()
+                parsed = WordVectors._parse_document(text)
+                documents.append(parsed)
+            except IOError:
+                continue
```
Comment (cclauss) [inline natural_language_processing/word_embeddings/word_embeddings.py:173]: Please put lines 166 to 173 in a separate function and then do
documents = [get_document_from_file(filename) for filename in glob.glob(full_path)]

File: requirements.txt [hunk 19]
```diff
@@ -17,3 +17,5 @@ sklearn
 sympy
 tensorflow
 xgboost
```
Comment (cclauss) [inline requirements.txt:19]: Please insert requirements in alphabetical order to reduce the likelihood of duplicate entries.

**Pull request closed without merging**